Define how to expose version through the collection metadata API (#1028)
OpenTermsArchive/engine: decision-records/0004-data-api.md

Currently, the terms annotated by ToS;DR are currently taken from original web documents stored on the ToS;DR server database. Open Terms Archive provides a significantly higher quality source for terms, for example by cleaning up all content that is not related to the terms, combining documents scattered across several pages, or normalising PDF documents.

To allow the transition from historical ToS;DR's crawler and database to the publicly accessible OTA datasets, ToS;DR requires access to the content of the versions through an API. This RFC outlines the proposed solutions.

## Proposed solutions

### Solution A:

#### Base URL

`<collection host>/api/:version`

#### Endpoint

##### `GET /version/:serviceId/:termsType/:date.md`

Return the markdown content of the version valid for the given date

##### Parameters

| Parameter | Type   | Description            |
| --------- | ------ | ---------------------- |
| serviceId | URL-encoded string | The ID of the service |
| termsType | URL-encoded string | The name of terms type |
| date | URL-encoded ISO date string | The date for which you want the version. This date should not be in the futur |

> Note about the date:

Reviewer: ```suggestion
Note about the date:
- A full date and time is required, and not a simple date (such as `2023-10-24`), to avoid ambiguities on days where a version changed, and timezone differences between client and server.
```

- It is not required that this date match exactly the fetch date of a version. As versions are fetched at periodic interval the version that will be returned will corresond to the version valid for this date range.
- To have the latest version available, simple use the current date time as parameter.

##### Returns

- A raw markdown content when a version was found for this date
- A HTTP 404 for a date that is anterior of the first available version

> - A HTTP 400 Bad Request, if the date is in the futur

Reviewer: ```suggestion
- HTTP 400 Bad Request with JSON content `{"error": "Requested version is in the future"}` if the date is in the future.
```

##### Example

```
GET /version/Open%20Terms%20Archive/Privacy%20Policy/2023-10-24T14%3A41%3A42.381Z.md
```
```markdown

> Privacy Policy

Reviewer: ```suggestion
HTTP 200
- - -
Privacy Policy
```

==============

Last updated: September 26, 2023

This privacy policy explains how, why and under what conditions the Open Terms Archive (OTA) site collects personal information and how it is used. Our privacy policy will change over time. And of course, we also record the changes of [our own documents](https://github.com/OpenTermsArchive/demo-versions/tree/main/Open%20Terms%20Archive) 😉

…
```

### Solution B:

#### Base URL

`<collection host>/api/:version`

#### Endpoint

##### `GET /version/:serviceId/:termsType/:date`

> Return a JSON object containing of the version content valid for the given date along some metadata

Reviewer: ```suggestion
Returns a JSON object containing the version content valid for the given date along with metadata.
```

##### Parameters

| Parameter | Type   | Description            |
| --------- | ------ | ---------------------- |
| serviceId | URL-encoded string | The ID of the service |
| termsType | URL-encoded string | The name of terms type |

> | date | URL-encoded ISO date string | The date for which you want the version. This date should not be in the futur |

Reviewer: | date | URL-encoded ISO 8601 datetime string | The date and time for which the version is requested |

> Note about the date:

Reviewer: ```suggestion
Note about the date:
- A full date and time is required, and not a simple date (such as `2023-10-24`), to avoid ambiguities on days where a version changed, and timezone differences between client and server.
```

> - It is not required that this date match exactly the fetch date of a version. As versions are fetched at periodic interval the version that will be returned will corresond to the version valid for this date range.

Reviewer: ```suggestion
- It is not required that this date matches exactly the fetch date of a version. As versions are fetched at a periodic interval, the version that will be returned will be the one that was applicable at the provided date.
```

> - To have the latest version available, simple use the current date time as parameter.

Reviewer: ```suggestion
- To get the latest version available, simply use the current date as parameter.
```

##### Returns

> - A raw markdown content when a version was found for this date

Reviewer: ```suggestion
- HTTP 200 with a JSON object containing the Markdown content as body when a version is found for this date
```

> - A HTTP 404 for a date that is anterior of the first available version

Reviewer: ```suggestion
- HTTP 404 Not Found with JSON content `{"error": "No version found for date <requested-date>"}` for a date that is anterior to the first available version
```

> - A HTTP 400 Bad Request, if the date is in the futur

Reviewer: ```suggestion
- HTTP 400 Bad Request with JSON content `{"error": "Requested version is in the future"}` if the date is in the future.
```

##### Example

```
GET /version/Open%20Terms%20Archive/Privacy%20Policy/2023-10-24T14%3A41%3A42.381Z
```
```json
{
  "fetchDate": "2023-10-20T06:30:00.381Z",
  "id": "c0dac2866fb2cdef7f8b98cc260177ac43df273b",
  "content": "
Privacy Policy
==============

Last updated: September 26, 2023

This privacy policy explains how, why and under what conditions the Open Terms Archive (OTA) site collects personal information and how it is used. Our privacy policy will change over time. And of course, we also record the changes of [our own documents](https://github.com/OpenTermsArchive/demo-versions/tree/main/Open%20Terms%20Archive) 😉

…

> "

Reviewer: This multi-line example is invalid. We would need to transform newlines to `\n`. The example should also provide an example with JSON data, to demonstrate how it would be encoded.

> # Define how to expose version through the collection metadata API

Reviewer: ```suggestion
# Expose versions data through the collection API
```

To allow the transition from historical ToS;DR's crawler and database to the publicly accessible Open Terms Archive collections, ToS;DR requires access to the content of the versions. Datasets are not satisfactory as they are produced on a weekly basis, whereas a workflow supporting annotation of just-added terms is required by ToS;DR. This RFC outlines possible implementations of an API providing this data.

## Proposed solutions

### Solution A

#### Base URL

`<collection host>/api/:version`

#### Endpoint

##### `GET /version/:serviceId/:termsType/:date.md`

Return the Markdown content of the version of the terms that was applicable at the given date

##### Parameters

| Parameter | Type   | Description            |
| --------- | ------ | ---------------------- |
| serviceId | URL-encoded string | The ID of the service |
| termsType | URL-encoded string | The name of terms type |

> | date | URL-encoded ISO 8601 datetime string | The date and time for which the version is requested |

Reviewer 2: And require that the timezone be UTC?

Note about the date:

> - A full date and time is required, and not a simple date (such as `2023-10-24`), to avoid ambiguities on days where a version changed, and timezone differences between client and server.

Reviewer 2: Maybe pin the timezone to UTC? So "A full UTC data and time"?

Author: If the timezone is defined as an offset from UTC (ex: `2023-12-06T12:34:56+02:00`) is still ISO 8601 valid. 
So, why forcing the timezone to be UTC?

}
```

</details>

#### Endpoint with Markdown content-type

##### `GET /version/:serviceId/:termsType/:date.md`

Return the contents of the version of the terms that was applicable at the given date in Markdown format.

##### Parameters

| Parameter | Type   | Description            |
| --------- | ------ | ---------------------- |
| serviceId | URL-encoded string | The ID of the service |
| termsType | URL-encoded string | The name of terms type |
| date | URL-encoded ISO 8601 datetime string | The date and time for which the version is requested |

Note about the date:
- A full date and time is required, and not a simple date (such as `2023-10-24`), to avoid ambiguities on days where a version changed, and timezone differences between client and server.
- To get the latest version available, simply use the current date as parameter.

##### Returns

- HTTP 200 with the Markdown content as body when a version is applicable at this date
- HTTP 404 Not Found with Markdown content `# Error\n\n_No version found for date <requested-date>_` for a date that is anterior to the first available version

> - HTTP 400 Bad Request with JSON content `# Error\n\n_Requested date <requested-date> is in the future, no version can exist there_` if the date is in the future.

Reviewer: There is an inconsistency with the Markdown / JSON content in this proposal here.
```suggestion
- HTTP 400 Bad Request with Markdown content `# Error\n\n_Requested date <requested-date> is in the future, no version can exist there_` if the date is in the future.
```

> - HTTP 400 Bad Request with JSON content `# Error\n\n_Requested date <requested-date> is in the future, no version can exist there_` if the date is in the future.

Reviewer: I suggest to use 416 instead of 400 for dates in the future, and to handle invalid dates with 400.

```suggestion
- HTTP 400 Bad Request with Markdown content `# Error\n\n_Requested date <requested-date> is not a valid ISO 8601 date time_` if the date is not a valid ISO 8601 date time.
- HTTP 416 Range Not Satisfiable with Markdown content `# Error\n\n_Requested date <requested-date> is in the future, no version can exist there_` if the date is in the future.
```
